### ISSUE TITLE
[core/lint] what should the returned data structure of `lint` look like?

### ISSUE THREAD

**Issue by samuelstroschein** [inlang/inlang#1098]
## Context

We need to change the return type of `lint`.

The currently returned data structure of `lint` returns `Result<[LintedResource[], Error[]>`. #1096 entails that no resources should be returned.

## Proposal 

We can take different directions here. 

| Proposal                      | Ease of incremental improvements | Ease of use in apps, etc. | Certainty of future-proofing |
|-------------------------------|----------------------------------|---------------------------|------------------------------|
| Result<LintReport[], Error[]> | medium                           | easy   | medium                       |
| Result<LintTree, Error[]>     | hard                             | hard (requires query)   | low                          |
| Result<Object, Error[]> | easy                             | easy | high                         |

### Proposal 1 - `Result<LintReport[], Error[]>`

```ts
const [lints, errors] = lint(config)

// good DX: makes it very easy to use filter
lints.filter(report => report.ruleId === "inlang.missingMessages")
lints.filter(report => report.messageId === "id.3")
lints.filter(report => report.messageId === "id.3" && report.resourceId === "en-DE")
```

#### Pros

- easy to implement in core
- "neutral" data structure (an array) which should make incremental updates easier because `LintReport` is the only driver of change
- better DX: removes need to separately import query functions

#### Cons

- changing the return type is a breaking change in the future because it's no object
- not performance optimized (But, the performance seems okay)

### Proposal 2 - `Result<LintTree, Error[]>`

_From https://github.com/inlang/inlang/issues/1074#issuecomment-1633629916_
> Agree, building a tree parallel to the Resources would make sense e.g.
> 
> ```ts
> const lintReport = {
> 	de: {
> 		'some.message': [
> 			{
> 				id: 'inlang.missingMessage',
> 				level: 'error',
> 				message: `Message with id 'some.message' is missing for 'de'.`
> 			},
> 			// ...
> 		]
> 	},
> 	// ...
> }
> ```

#### Pros

- (prematurely) performance optimized (proposal 3 can use this data structure under the hood)

#### Cons

- leaks implementation to the ecosystem which makes refactoring hard
- requires separate query functions
- likely not future-proof because we don't know yet how inlang will evolve and if the chosen tree structure is a good fit in the future

### Proposal 3 - `Result<Object, Error[]>`

```ts
type Lints = {
  // the used rules for easy showing of "those are the rules"
  rules: LintRule[]
  // all lint reports for manual filtering
  all: LintReport[] 
  // future... we can extend the object however we need
}
```

```ts

const [lints, errorsDuringLint] = lint(config) 

lints.rules
>> LintRule[]

// only for message
lints.get()
>> Returns all lint reports as an array (essentially proposal 1)

// same easy API as proposal 1 can be used
lints.get().filter(report => report.ruleId === "inlang.missingMessages")
lints.get().filter(report => report.messageId === "id.3")

// IN THE FUTURE (if performance becomes critical), custom performance-optimized filters can be easily introduced. 
// lints.get({ messageId: "id.3" })
``` 

#### Pros

- future-proofer: the returned object can be extended as we need 
- future-proofer: the `get()` function can be extended and optimized in the future
- doesn't like performance optimizations into the ecosystem like proposal 2

#### Cons

- ???

**Comment by samuelstroschein** [comment]
I vote for proposal 3 because it is future-proof(er) and doesn't require query building at the moment.

**Comment by samuelstroschein** [comment]
Proposal 2

**Comment by samuelstroschein** [comment]
Proposal 3

**Comment by samuelstroschein** [comment]
@inlang/editor you worked with lint filters, please vote/give feedback. 

@inlang/ide-extension you also worked with lint filters, please vote/give feedback. 

@ivanhofer your input is desired too

**Comment by samuelstroschein** [comment]
@ivanhofer why did you vote for proposal 2? 

It seems to be the worst out of all proposals. Please don't reply with performance reasons because proposal 3 can easily achieve the same performance.

**Comment by ivanhofer** [comment]
I think a dedicated utility to iterate over lint results is more flexible than solution 3. This would open up the possibilities for other projects to easier work with the raw results instead of using the provided query functions.
We still can introduce solution 3 in the future if we see that it makes more sense (adding something is easier than removing a functionality).

Solution 1 or 2 are similar. Solution 1 needs to have more props attached to a single result l, where in solution 2 it is integrated into the tree structure.
I can't remember the implementation details, but I think we can easily provide the language information and message id automatically, without the need of the caller of `report()` to set them. In this case I would vote for solution 1, with a dedicated (external) query tool.

**Comment by ivanhofer** [comment]
> @ivanhofer why did you vote for proposal 2? 
> 
> It seems to be the worst out of all proposals. Please don't reply with performance reasons because proposal 3 can easily achieve the same performance. 

GitHub on mobile is buggy 😅 I have fixed my votes

**Comment by samuelstroschein** [comment]
@ivanhofer I updated proposal 3 because query will not be included for now. Please read the pros and cons!

> I think a dedicated utility to iterate over lint results is more flexible than solution 3. This would open up the possibilities for other projects to easier work with the raw results instead of using the provided query functions.

But... `lints.get()` returns an array for external people to use.

> We still can introduce solution 3 in the future if we see that it makes more sense (adding something is easier than removing a functionality).

No, we can't. 

If proposal 1 or 2 are chosen, we can never extend the return type because it's no object. Proposal 3 specifically targets future incremental improvements.

**Comment by ivanhofer** [comment]
Ok, if serialization is not a use case then we can go for solution 3. I don't think we need to serialize the result, and if yes this would also just be a small overhead of accessing all lints first.

**Comment by felixhaeberle** [comment]
I second Proposal 3 but would change the `all` property to `report`. Because `LintRule[]` -> `rules`, `LintReport[]` -> `report`.

Or did I miss something and the `all` makes more sense?

**Comment by samuelstroschein** [comment]
> Or did I miss something and the all makes more sense? 

I agree. The correct name would be `reports` though. Clashes a tiny bit with `lints.reports` but better than `all` for sure.

**Comment by samuelstroschein** [comment]
If we decide on `lints.reports`, wouldn't it make sense to put `exceptions` on lints as well?

Instead of

```ts
const [lints, exceptions] = lint(messages)

if (exceptions){
  // do something
}

lints.reports
```

Proposal

- no `result` type which makes lints easier to understand?

```ts
const lints = lint(messages)

if (lints.exceptions) {
  // do something
}
lints.reports


```

**Comment by felixhaeberle** [comment]
Yes please, no separate exception bundling; this only makes it harder to assign it back again to the specific message for outputting it to the user.

**Comment by ivanhofer** [comment]
> If we decide on `lints.reports`, wouldn't it make sense to put `exceptions` on lints as well?

Nobody will check for `exceptions` because it is hidden. We will have some upcoming bugs if we decide to go that route..

**Comment by ivanhofer** [comment]
> That's the case now too.

https://github.com/ivanhofer/exceptionally 👀 ^^

**Comment by samuelstroschein** [comment]
will be addressed in #1149.
